File: android/src/main/java/com/cas/MediationManagerModule.kt
package com.cas

import com.cas.extensions.MediationManagerWrapper
import com.cas.extensions.fromReadableMap
import com.cas.extensions.toInt
import com.cas.extensions.toReadableMap
import com.cleversolutions.ads.*
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule


class MediationManagerModule(context: ReactApplicationContext, private val managerWrapper: MediationManagerWrapper): ReactContextBaseJavaModule(context), AdLoadCallback,
  AdCallback {
  override fun getName() = "MediationManagerModule"
  private val subId = managerWrapper.addLoadListener(this)
  private val changeId = managerWrapper.addChangeListener { it?.let { initCASAppOpen() } }
  private var appOpenAd: CASAppOpen? = null

  companion object {
    private val APP_OPEN_AD_TYPE = 5
  }

  override fun invalidate() {
    managerWrapper.removeLoadListener(subId)
    managerWrapper.removeChangeListener(changeId)
    super.invalidate()
  }

  private fun initCASAppOpen() {
    appOpenAd = CASAppOpen.create(managerWrapper.manager!!)
  }

  @ReactMethod
  @Suppress("unused")
  fun setLastPageAdContent(content: ReadableMap) {
    managerWrapper.manager?.let {
      it.lastPageAdContent = LastPageAdContent::class.fromReadableMap(content)
    }
  }

  /*
  Interstitial
   */
  @ReactMethod
  @Suppress("unused")
  fun loadInterstitial() {
    managerWrapper.manager?.loadInterstitial()
  }

  @ReactMethod
  @Suppress("unused")
  fun isInterstitialReady(promise: Promise) {
    promise.resolve(managerWrapper.manager?.isInterstitialReady)
  }

  @ReactMethod
  @Suppress("unused")
  fun showInterstitial(callbackId: String) {
    reactApplicationContext.currentActivity?.let {
      managerWrapper.manager?.showInterstitial(it, createJSAdCallback("Interstitial", callbackId))
    }
  }

  /*
  RewardedAd
   */
  @ReactMethod
  @Suppress("unused")
  fun loadRewardedAd() {
    managerWrapper.manager?.loadRewardedAd()
  }

  @ReactMethod
  @Suppress("unused")
  fun isRewardedAdReady(promise: Promise) {
    promise.resolve(managerWrapper.manager?.isRewardedAdReady)
  }

  @ReactMethod
  @Suppress("unused")
  fun showRewardedAd(callbackId: String) {
    reactApplicationContext.currentActivity?.let {
      managerWrapper.manager?.showRewardedAd(it, createJSAdCallback("Rewarded", callbackId))
    }
  }

  /*
  AppReturn
   */
  @ReactMethod
  @Suppress("unused")
  fun enableAppReturnAds(callbackId: String) {
    managerWrapper.manager?.enableAppReturnAds(createJSAdCallback("AppReturn", callbackId))
  }

  @ReactMethod
  @Suppress("unused")
  fun disableAppReturnAds() {
    managerWrapper.manager?.disableAppReturnAds()
  }

  @ReactMethod
  @Suppress("unused")
  fun skipNextAppReturnAds() {
    managerWrapper.manager?.skipNextAppReturnAds()
  }

  /*
  AppOpen
   */
  @ReactMethod
  @Suppress("unused")
  fun loadAppOpenAd(isLandscape: Boolean) {
    appOpenAd?.loadAd(reactApplicationContext, isLandscape, object : LoadAdCallback {
        override fun onAdFailedToLoad(error: AdError) {
          val map = WritableNativeMap()

          map.putInt("type", APP_OPEN_AD_TYPE)
          map.putString("error", error.message)

          reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("adFailedToLoad", map)
        }

        override fun onAdLoaded() {
          val map = WritableNativeMap()

          map.putInt("type", APP_OPEN_AD_TYPE)

          reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("adLoaded", map)
        }
    })
  }

  @ReactMethod
  @Suppress("unused")
  fun isAppOpenAdAvailable(promise: Promise) {
    promise.resolve(appOpenAd?.isAdAvailable())
  }

  @ReactMethod
  @Suppress("unused")
  fun showAppOpenAd(callbackId: String) {
    reactApplicationContext.currentActivity?.let { activity ->
      appOpenAd?.let { ad ->
        ad.contentCallback = createJSAdCallback("AppOpen", callbackId)
        ad.show(activity)
      }
    }
  }

  private fun createJSAdCallback(module: String, callbackId: String): AdPaidCallback {
    return object : AdPaidCallback {
      override fun onShown(ad: AdImpression) {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)
        //map.putMap("data", ad.toReadableMap())

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onShown", map)
      }

      override fun onShowFailed(message: String) {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)
        map.putString("data", message)

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onShowFailed", map)
      }

      override fun onAdRevenuePaid(ad: AdImpression) {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)
        map.putMap("data", ad.toReadableMap())

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onAdRevenuePaid", map)
      }

      override fun onClicked() {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onClicked", map)
      }

      override fun onComplete() {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onComplete", map)
      }

      override fun onClosed() {
        val map = WritableNativeMap()
        map.putString("callbackId", callbackId)

        reactApplicationContext
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
          .emit("onClosed", map)
      }
    }
  }

  override fun onAdFailedToLoad(type: AdType, error: String?) {
    val map = WritableNativeMap()

    map.putInt("type", type.toInt())
    map.putString("error", error)

    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("adFailedToLoad", map)
  }

  override fun onAdLoaded(type: AdType) {
    val map = WritableNativeMap()

    map.putInt("type", type.toInt())

    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("adLoaded", map)
  }

  /*
  RN stubs
   */
  @ReactMethod
  fun addListener(eventName: String?) {}

  @ReactMethod
  fun removeListeners(count: Int?) {}
}

File: ios/MediationManagerModule.swift
//
//  MediationManagerModule.swift
//  react-native-cas
//
//  Created by Lonely Astronaut on 20.09.23.
//

import Foundation
import CleverAdsSolutions

@objc(MediationManagerModule)
class MediationManagerModule: RCTEventEmitter, CASLoadDelegate {
    private var subId = ""
    private var changeId = ""
    private var appOpenAd: CASAppOpen? = nil
    private var callbacks: Dictionary<String, JSAdCallback> = [:]
    private static let APP_OPEN_AD_TYPE = 3
    
    override init() {
        super.init()
        subId = MediationManagerWrapper.shared.addLoadListener(self)
        changeId = MediationManagerWrapper.shared.addChangeListener { self.initCASAppOpen(manager: $0) }
    }
    
    override func supportedEvents() -> [String]! {
        ["onShown", "onAdRevenuePaid", "onShowFailed", "onClicked", "onComplete", "onClosed", "adLoaded", "adFailedToLoad"]
    }
    
    override class func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    func initCASAppOpen(manager: CASMediationManager) {
        appOpenAd = CASAppOpen.create(manager: manager)
    }
    
    @objc func setLastPageAdContent(_ content: NSDictionary) {
        MediationManagerWrapper.shared.manager?.lastPageAdContent = CASLastPageAdContent.fromDictionary(jsMap: content)
    }
    
    @objc func loadInterstitial() {
        DispatchQueue.main.async {
            MediationManagerWrapper.shared.manager?.loadInterstitial()
        }
    }
    
    @objc func isInterstitialReady(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(MediationManagerWrapper.shared.manager?.isInterstitialReady)
    }
    
    @objc func showInterstitial(_ callbackId: NSString) {
        DispatchQueue.main.async {
            MediationManagerWrapper.shared.manager?.presentInterstitial(
                fromRootViewController: UIApplication.shared.delegate!.window!!.rootViewController!,
                callback: self.createJSAdCallback("Interstitial", callbackId)
            )
        }
    }
    
    @objc func loadRewardedAd() {
        MediationManagerWrapper.shared.manager?.loadRewardedAd()
    }
    
    @objc func isRewardedAdReady(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(MediationManagerWrapper.shared.manager?.isRewardedAdReady)
    }
    
    @objc func showRewardedAd(_ callbackId: NSString) {
        MediationManagerWrapper.shared.manager?.presentRewardedAd(
            fromRootViewController: UIApplication.shared.delegate!.window!!.rootViewController!,
            callback: createJSAdCallback("Rewarded", callbackId)
        )
    }
    
    @objc func enableAppReturnAds(_ callbackId: NSString) {
        let key = UUID().uuidString;
        let cb = JSAppReturn(callbackId: callbackId as String) { [weak self] name, body in
            self?.sendEvent(withName: name, body: body)
            
            // cleanup memory
            if name == "onClosed" || name == "onShowFailed" {
                self?.callbacks.removeValue(forKey: key)
            }
        }
        
        callbacks[key] = cb
        
        MediationManagerWrapper.shared.manager?.enableAppReturnAds(with: cb)
    }
    
    @objc func disableAppReturnAds() {
        MediationManagerWrapper.shared.manager?.disableAppReturnAds()
    }
    
    @objc func skipNextAppReturnAds() {
        MediationManagerWrapper.shared.manager?.skipNextAppReturnAds()
    }
    
    @objc func loadAppOpenAd(_ isLandscape: Bool) {
        appOpenAd?.loadAd() {[weak self] ad, error in
            if let error = error {
                self?.sendEvent(withName: "adFailedToLoad", body: [
                    "type": MediationManagerModule.APP_OPEN_AD_TYPE,
                    "error": error.localizedDescription
                ])
            } else {
                self?.sendEvent(withName: "adLoaded", body: [
                    "type": MediationManagerModule.APP_OPEN_AD_TYPE
                ])
            }
        }
    }
    
    @objc func isAppOpenAdAvailable(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(appOpenAd?.isAdAvailable())
    }
    
    @objc func showAppOpenAd(_ callbackId: NSString) {
        appOpenAd?.contentCallback = createJSAdCallback("AppOpen", callbackId)
        appOpenAd?.present(fromRootViewController: UIApplication.shared.delegate!.window!!.rootViewController!)
    }
    
    private func createJSAdCallback(_ module: String, _ callbackId: NSString) -> CASCallback {
        let key = UUID().uuidString
        let callback = JSAdCallback(callbackId: callbackId as String) { [weak self] name, body in
            self?.sendEvent(withName: name, body: body)
            
            // cleanup memory
            if name == "onClosed" || name == "onShowFailed" {
                self?.callbacks.removeValue(forKey: key)
            }
        }
        
        callbacks[key] = callback
        
        return callback
    }
    
    func onAdLoaded(_ adType: CleverAdsSolutions.CASType) {
        sendEvent(withName: "adLoaded", body: [
            "type": adType.rawValue
        ])
    }
    
    func onAdFailedToLoad(_ adType: CleverAdsSolutions.CASType, withError error: String?) {
        sendEvent(withName: "adFailedToLoad", body: [
            "type": adType.rawValue,
            "error": error as Any
        ])
    }
    
    override func invalidate() {
        MediationManagerWrapper.shared.removeLoadListener(subId)
        MediationManagerWrapper.shared.removeChangeListener(changeId)
        super.invalidate()
    }
}

class JSAdCallback: CASPaidCallback {
    private let callbackId: String
    private let sendEvent: (_ eventName: String, _ body: Any) -> Void
    
    init(callbackId: String, sendEvent: @escaping (_ eventName: String, _ body: Any) -> Void) {
        self.callbackId = callbackId
        self.sendEvent = sendEvent
    }
    
    func didPayRevenue(for ad: CleverAdsSolutions.CASImpression) {
        sendEvent("onAdRevenuePaid", [
            "callbackId": callbackId,
            "data": ad.toDictionary() as Any
        ])
    }
    
    func willShown(ad adStatus: CASImpression) {
        sendEvent("onShown", [
            "callbackId": callbackId,
            "data": adStatus.toDictionary() as Any
        ])
    }
    
    func didClosedAd() {
        sendEvent("onClosed", [
            "callbackId": callbackId,
        ])
    }
    
    func didClickedAd() {
        sendEvent("onClicked", [
            "callbackId": callbackId,
        ])
    }
    
    func didCompletedAd() {
        sendEvent("onComplete", [
            "callbackId": callbackId,
        ])
    }
    
    func didShowAdFailed(error: String) {
        sendEvent("onShowFailed", [
            "callbackId": callbackId,
            "data": error
        ])
    }
}

class JSAppReturn: JSAdCallback, CASAppReturnDelegate {
    func viewControllerForPresentingAppReturnAd() -> UIViewController {
        return UIApplication.shared.delegate!.window!!.rootViewController!
    }
}
